packages/grid/_modules_/grid/hooks/features/columnPinning/useGridColumnPinning.tsx

```diff
@@ -0,0 +1,310 @@
+      if (nbLeftPinnedColumns + nbRightPinnedColumns === 0) {
+        return;
+      }
+
+      const index = event.currentTarget.dataset.rowindex;
+      const rowElements = apiRef.current.windowRef!.current!.querySelectorAll(
+        `.${gridClasses.row}[data-rowindex="${index}"]`,
+      );
+      rowElements.forEach((row) => {
+        if (event.type === 'mouseenter') {
+          row.classList.add('Mui-hovered');
+        } else {
+          row.classList.remove('Mui-hovered');
+        }
+      });
+    },
+    [apiRef, pinnedColumns.left, pinnedColumns.right, props.disableColumnPinning],
+  );
+
+  const handleMouseEnter = React.useCallback(
+    (params: GridRowParams, event: MuiEvent<React.MouseEvent<HTMLElement>>) => {
+      updateHoveredClassOnSiblingRows(event);
+    },
+    [updateHoveredClassOnSiblingRows],
+  );
+
+  const handleMouseLeave = React.useCallback(
+    (params: GridRowParams, event: MuiEvent<React.MouseEvent<HTMLElement>>) => {
+      updateHoveredClassOnSiblingRows(event);
+    },
+    [updateHoveredClassOnSiblingRows],
+  );
+
+  useGridApiEventHandler(apiRef, GridEvents.rowMouseEnter, handleMouseEnter);
+  useGridApiEventHandler(apiRef, GridEvents.rowMouseLeave, handleMouseLeave);
+
+  const calculateScrollLeft = React.useCallback(
+    (initialValue, params: Partial<GridCellIndexCoordinates>) => {
+      if (props.disableColumnPinning) {
+        return initialValue;
+      }
+
+      const visibleColumns = visibleGridColumnsSelector(apiRef.current.state);
+      const [leftPinnedColumns, rightPinnedColumns] = filterColumns(pinnedColumns, visibleColumns);
+
+      if (!params.colIndex || (leftPinnedColumns.length === 0 && rightPinnedColumns.length === 0)) {
+        return initialValue;
+      }
+
+      const columnsMeta = gridColumnsMetaSelector(apiRef.current.state);
+      const clientWidth = apiRef.current.windowRef!.current!.clientWidth;
+      const scrollLeft = apiRef.current.windowRef!.current!.scrollLeft;
+      const offsetWidth = visibleColumns[params.colIndex].computedWidth;
+      const offsetLeft = columnsMeta.positions[params.colIndex];
+
+      const leftPinnedColumnsWidth = columnsMeta.positions[leftPinnedColumns.length];
+      const rightPinnedColumnsWidth =
+        columnsMeta.totalWidth -
+        columnsMeta.positions[columnsMeta.positions.length - rightPinnedColumns.length];
+
+      const elementBottom = offsetLeft + offsetWidth;
+      if (elementBottom - (clientWidth - rightPinnedColumnsWidth) > scrollLeft) {
+        const left = elementBottom - (clientWidth - rightPinnedColumnsWidth);
+        return { ...initialValue, left };
+      }
+      if (offsetLeft < scrollLeft + leftPinnedColumnsWidth) {
+        const left = offsetLeft - leftPinnedColumnsWidth;
+        return { ...initialValue, left };
+      }
+      return initialValue;
+    },
+    [apiRef, pinnedColumns, props.disableColumnPinning],
+  );
+
+  const addColumnMenuButtons = React.useCallback(
+    (initialValue) => {
```

Comment:
```suggestion
    (initialValue: JSX.Element[]) => {
```

Could we type `initialValue` here ?
I'll do a PR to correctly type the pre-processors one day :+1:

Reply:
Done.